oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/authentication/OpaqueTokenAuthenticationProvider.java

```diff
@@ -0,0 +1,256 @@
+		Assert.notNull(introspectionUri, "introspectionUri cannot be null");
+		Assert.notNull(clientId, "clientId cannot be null");
+		Assert.notNull(clientSecret, "clientSecret cannot be null");
+
+		this.introspectionUri = URI.create(introspectionUri);
+		RestTemplate restTemplate = new RestTemplate();
+		restTemplate.getInterceptors().add(new BasicAuthenticationInterceptor(clientId, clientSecret));
+		this.restOperations = restTemplate;
+	}
+
+	/**
+	 * Creates a {@code OpaqueTokenAuthenticationProvider} with the provided parameters
+	 *
+	 * @param introspectionUri The introspection endpoint uri
+	 * @param restOperations The client for performing the introspection request
+	 */
+	public OpaqueTokenAuthenticationProvider(String introspectionUri, RestOperations restOperations) {
+		Assert.notNull(introspectionUri, "introspectionUri cannot be null");
+		Assert.notNull(restOperations, "restOperations cannot be null");
+
+		this.introspectionUri = URI.create(introspectionUri);
+		this.restOperations = restOperations;
+	}
+
+	/**
+	 * Introspect and validate the opaque
+	 * <a href="https://tools.ietf.org/html/rfc6750#section-1.2" target="_blank">Bearer Token</a>.
+	 *
+	 * @param authentication the authentication request object.
+	 *
+	 * @return A successful authentication
+	 * @throws AuthenticationException if authentication failed for some reason
+	 */
+	@Override
+	public Authentication authenticate(Authentication authentication) throws AuthenticationException {
+		if (!(authentication instanceof BearerTokenAuthenticationToken)) {
+			return null;
+		}
+
+		// introspect
+		BearerTokenAuthenticationToken bearer = (BearerTokenAuthenticationToken) authentication;
+		TokenIntrospectionSuccessResponse response = introspect(bearer.getToken());
+		Map<String, Object> claims = extractClaimSet(response);
+		Instant iat = (Instant) claims.get("iat");
+		Instant exp = (Instant) claims.get("exp");
+
+		// construct token
+		OAuth2TokenWithClaims withClaims = new OAuth2TokenWithClaims(bearer.getToken(), iat, exp, claims);
+		Collection<GrantedAuthority> authorities = extractAuthorities(claims);
+		AbstractAuthenticationToken result = new IntrospectedOAuth2TokenAuthenticationToken(withClaims, authorities);
+		result.setDetails(bearer.getDetails());
+		return result;
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	@Override
+	public boolean supports(Class<?> authentication) {
+		return BearerTokenAuthenticationToken.class.isAssignableFrom(authentication);
+	}
+
+	private TokenIntrospectionSuccessResponse introspect(String token) {
+		HttpHeaders headers = new HttpHeaders();
+		headers.setAccept(Collections.singletonList(MediaType.APPLICATION_JSON_UTF8));
+		RequestEntity<String> requestEntity =
+				new RequestEntity<>("token=" + token, headers, HttpMethod.POST, this.introspectionUri);
+		TokenIntrospectionSuccessResponse response = asSuccess(parse(adapt(request(requestEntity))));
```

Comment:
I don't see any benefit in using Nimbus' API's to extract the response. The response returned is really a `Map` that can easily be handled by `RestTemplate`. Or another option and the one I prefer is to provide a `HttpMessageConverter`. See `OAuth2AccessTokenResponseHttpMessageConverter` which is used by `DefaultAuthorizationCodeTokenResponseClient`.

Reply:
The benefit is that we don't need to write our own `HttpMessageConverter`. Why do you think we should write this ourselves?

Reply:
It just seems inconsistent to use `RestTemplate` to initiate the request and Nimbus API to parse the response. `HttpMessageConverter` is part of `RestTemplate`'s API so we should leverage it for response parsing.

As well, the main benefit is that the user will be able to customize the response processing by either extending our implementation of `HttpMessageConverter` for the Introspection Response or providing their own. Keep in mind that custom parameters may come back from the Introspection Response so users will need a way to access them.

Reply:
@rwinch When you get a chance can you provide feedback here.

Reply:
I spoke to @rwinch about this one and he also felt that being able to customize the response can be left out for now. As well, using Nimbus  API for response processing is consistent with what `ClientRegistrations` does today.